### PR TITLE
Refactored chaindatafetcher

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Modified `Repository` interface to reduce the number of insertion methods
- Defined `HandleChainEventFn` function for code readability
- Created `types` package to avoid cyclic dependency
- Refactored chaindatafetcher `handleRequestByType`

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/kas/repository.go [hunk 1]
```diff
@@ -98,6 +100,21 @@ func (r *repository) SetComponent(component interface{}) {
...
+	case types.RequestTypeTransaction:
+		return r.InsertTransactions(event)
+	case types.RequestTypeTrace:
+		return r.InsertTraceResults(event)
+	case types.RequestTypeTokenTransfer:
+		return r.InsertTokenTransfers(event)
+	case types.RequestTypeContract:
+		return r.InsertContracts(event)
+	default:
+		return fmt.Errorf("unsupported data type: %v", reqType)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/kas/repository.go:114]: How about also leaving the information from ChainEvent?

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 2]
```diff
@@ -354,29 +353,75 @@ func (f *ChainDataFetcher) updateCheckpoint(num int64) error {
...
+	case cfTypes.RequestTypeTransaction:
+		return txsInsertionTimeGauge, nil
+	case cfTypes.RequestTypeTokenTransfer:
+		return tokenTransfersInsertionTimeGauge, nil
+	case cfTypes.RequestTypeContract:
+		return contractsInsertionTimeGauge, nil
+	case cfTypes.RequestTypeTrace:
+		return tracesInsertionTimeGauge, nil
+	default:
+		return nil, fmt.Errorf("the request type is not supported. type: %v", reqType)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:367]: If you leave a warning log and return an empty gauge, you don't need error handling logic at updateInsertionTimeGauge.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 3]
```diff
@@ -354,29 +353,75 @@ func (f *ChainDataFetcher) updateCheckpoint(num int64) error {
...
+		} else {
+			gauge.Update(elapsed.Milliseconds())
+		}
+		return nil
 	}
 }
 
-func (f *ChainDataFetcher) retryFunc(insert func(blockchain.ChainEvent) error, gauge metrics.Gauge) func(blockchain.ChainEvent) {
-	return func(event blockchain.ChainEvent) {
+func getInsertionRetryGauge(reqType cfTypes.RequestType) (metrics.Gauge, error) {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:389]: Same comment for error handling logic.

**Review by aidan-kwon — APPROVED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Agreed. Reflected on a9b659884694577dba20dad99f4c46673df1aab3.

File: datasync/chaindatafetcher/kas/repository.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/kas/repository.go]: Added block number to the error. 1c477b9014671902ffff40079ea39a2142812750

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Agreed. Reflected on a9b659884694577dba20dad99f4c46673df1aab3.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 4]
```diff
@@ -353,18 +353,19 @@ func (f *ChainDataFetcher) updateCheckpoint(num int64) error {
...
 	case cfTypes.RequestTypeContract:
-		return contractsInsertionTimeGauge, nil
+		return contractsInsertionTimeGauge
 	case cfTypes.RequestTypeTrace:
-		return tracesInsertionTimeGauge, nil
+		return tracesInsertionTimeGauge
 	default:
-		return nil, fmt.Errorf("the request type is not supported. type: %v", reqType)
+		logger.Warn("the request type is not supported", "type", reqType)
+		return metrics.NewGauge()
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:368]:
```suggestion
		return metrics.NilGauge{}
```

**Review by aidan-kwon — APPROVED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Thanks. Reflected on 2ea1febde0d4025ee16b712b4de5dd574d83c4cf.

File: kafka-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
+version: "3.7"
```
Comment (ehnuje) [inline kafka-compose.yaml:1]: Is this what you intended?

Comment (jeongkyun-oh) [reply kafka-compose.yaml]: My bad. I removed it.

**Review by aidan-kwon — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Pull request merged**